### DISCUSSION TITLE
FastAPI does not respect pydantic v2 model_config

### DISCUSSION THREAD

**Discussion by mikkelam** [fastapi/fastapi#10334] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from datetime import timedelta
from fastapi import FastAPI
import uvicorn
from pydantic import BaseModel, ConfigDict
app = FastAPI()

class A(BaseModel):
    model_config = ConfigDict(ser_json_timedelta="float")
    td: timedelta

@app.get("/")
async def root() -> A:
    a = A(td=timedelta(seconds=10))
    print(a.model_dump_json())
    return a

if __name__ == "__main__":
    uvicorn.run(app, port=1337)
```

### Description

Calling the above endpoint returns:

```json
{"td":"PT10S"}
```
However, the console prints:
```
{"td":10.0}
```

Which  is the correct and expected behavior

### Operating System

macOS

### FastAPI Version

0.103.1

### Pydantic Version

2.4.0

### Python Version

3.11.5

**Comment by jurasofish** [comment]
I've seen the same thing and worked around it with a custom `field_serializer`

```python
from datetime import timedelta
from fastapi import FastAPI
import uvicorn
from typing import Any
from pydantic import BaseModel, ConfigDict, field_serializer
import threading
app = FastAPI()

class A(BaseModel):
    model_config = ConfigDict(ser_json_timedelta="float")
    td: timedelta

    @field_serializer("td")
    def serialize_dt(self, index_freq: timedelta, _info: Any) -> float:
        return index_freq.total_seconds()

@app.get("/")
async def root() -> A:
    a = A(td=timedelta(seconds=10))
    print(a.model_dump_json())
    return a

def do_something():
    import requests
    import time
    time.sleep(1)  # wait for server to start xd
    print(requests.get("http://localhost:1337/").json())

if __name__ == "__main__":
    threading.Thread(target=do_something).start()
    uvicorn.run(app, port=1337)
```

**Comment by morian** [comment]
Hi !

I stumbled on the same issue and I think I found the root cause and a (very dirty) way to fix it.
The problem has already been described by @mikkelam, so I'll skip go straight to the dirty part:
```python
from datetime import timedelta
from fastapi import FastAPI, utils
from pydantic import BaseModel, ConfigDict, TypeAdapter

class Task(BaseModel):
    duration: timedelta

    model_config = ConfigDict(ser_json_timedelta='float')

def fixup_post_init(self) -> None:
    self._type_adapter = TypeAdapter(self.field_info.annotation)

def factory() -> FastAPI:
    utils.ModelField.__post_init__ = fixup_post_init

    app = FastAPI()

    @app.get('/')
    async def root_get() -> Task:
        return Task(duration='01:00:00')

    return app
```
There are probably other places during app initialization where this works, but this does for me on version 0.108.
The offending line is located here: https://github.com/tiangolo/fastapi/blob/0.108.0/fastapi/_compat.py#L106
```python
if PYDANTIC_V2:
    # [...]

    @dataclass
    class ModelField:
        # [...]

        def __post_init__(self) -> None:
            self._type_adapter: TypeAdapter[Any] = TypeAdapter(
                Annotated[self.field_info.annotation, self.field_info]
```
Any return type if wrapped in an annotation, itself passed in pydantic's `TypeAdapter`.


Now let's consider the following pydantic-only code:
 ```python
from datetime import timedelta
from typing import Annotated
from pydantic import BaseModel, ConfigDict, TypeAdapter

class Task(BaseModel):
    duration: timedelta

    model_config = ConfigDict(ser_json_timedelta='float')


d1 = Task(duration=timedelta(hours=1))
print('d1:', d1.model_dump_json())

PlainAdapter = TypeAdapter(Task)
d2 = PlainAdapter.validate_python(d1.model_dump())
print('d2:', PlainAdapter.dump_python(d2, mode='json'))

AnnotatedAdapter = TypeAdapter(Annotated[Task, None])
d3 = AnnotatedAdapter.validate_python(d1.model_dump())
print('d3:', AnnotatedAdapter.dump_python(d3, mode='json'))
```

When running this code, I get the following output (with pydantic v2.5.3):
```console
d1: {"duration":3600.0}
d2: {'duration': 3600.0}
d3: {'duration': 'PT3600S'}
```

From what I understand here, the `Annotated` wrapper breaks the configuration (or reset it to a default one).
This means that our own configuration is not applied, hence the dirty fix that removes this wrapper altogether.

On a side note, `TypeAdapter` takes an optional `config` keyword.

Maybe we could provide FastAPI with a default pydantic configuration that will be passed to this `TypeAdapter`,
or this method could determine automatically the configuration from the response model (if applicable).

**Reply by morian** [reply]
@Kludex , can someone have a look and tell whether this should be raised as an issue (in fastapi or pydantic)?

**Reply by Kludex** [reply, marked as answer]
It's a FastAPI issue.

**Comment by morian** [comment]
Just a note here, this was "fixed" pydantic in v2.6.

The temporary fix now raises an error with pydantic:
```
pydantic.errors.PydanticUserError: Cannot use `config` when the type is a BaseModel, dataclass or TypedDict. These types can have their own config and setting the config via the `config` parameter to TypeAdapter will not override it, thus the `config` you passed to TypeAdapter becomes meaningless, which is probably not what you want.
```

The pydantic-only code I provided above now works out of the box.

Related patches can be tracker through the following issue:
https://github.com/pydantic/pydantic/issues/8214
